solidity/contracts/AcreMultiAssetVault.sol

```diff
@@ -0,0 +1,284 @@
+    /// @dev This function can only be called by the owner.
+    /// @param asset Address of the asset to be added to the supported assets.
+    function addSupportedAsset(address asset) external onlyOwner {
+        return _addSupportedAsset(asset);
+    }
+
+    /// @notice Adds an asset to the list of assets supported by the vault.
+    /// @param asset Address of the asset to be added to the supported assets.
+    function _addSupportedAsset(address asset) internal {
+        if (asset == address(0)) {
+            revert ZeroAddress();
+        }
+        if (supportedAssets[asset]) {
+            revert AssetAlreadySupported();
+        }
+
+        emit SupportedAssetAdded(asset);
+
+        supportedAssets[asset] = true;
+    }
+
+    /// @notice Removes an asset from the list of assets supported by the vault.
+    /// @dev This function can only be called by the owner.
+    /// @param asset Address of the asset to be removed from the supported assets.
+    function removeSupportedAsset(address asset) external onlyOwner {
+        if (!supportedAssets[asset]) {
+            revert AssetNotSupported();
+        }
+
+        emit SupportedAssetRemoved(asset);
+
+        supportedAssets[asset] = false;
+    }
+
+    /// @notice Deposits an asset into the vault.
+    /// @param asset Address of the asset to be deposited.
+    /// @param amount Amount of the asset to be deposited (in the asset precision).
+    /// @return The ID of the deposit.
+    function deposit(address asset, uint256 amount) external returns (uint256) {
+        return depositFor(asset, amount, msg.sender);
+    }
+
+    /// @notice Deposits assets into the vault on behalf of a specified account,
+    ///         which will become the owner of the deposit. Only the owner of the
+    ///         deposit can withdraw the assets.
+    /// @param asset Address of the asset to be deposited.
+    /// @param amount The amount of deposit.
+    /// @param depositOwner The address of the account for which the deposit is
+    ///        being made.
+    /// @return The ID of the deposit.
+    function depositFor(
+        address asset,
+        uint256 amount,
+        address depositOwner
+    ) public returns (uint256) {
+        if (!supportedAssets[asset]) {
+            revert AssetNotSupported();
+        }
+        if (amount == 0) {
+            revert InvalidAmount(amount);
+        }
+        if (depositOwner == address(0) || depositOwner == asset) {
+            revert InvalidDepositOwner(depositOwner);
+        }
+
+        // Generate a new deposit ID.
+        uint256 depositId = ++depositCount;
+
+        emit DepositCreated(depositOwner, asset, depositId, amount);
+
+        // Transfer the asset from the deposit owner to the vault and approve
+        // Mezo Portal to pull them.
+        IERC20(asset).safeTransferFrom(msg.sender, address(this), amount);
+        IERC20(asset).forceApprove(address(mezoPortal), amount);
```

Comment:
There is a theoretical reentrancy risk here as we are interacting with a third-party address before making state changes. Sure, the asset must be supported here and in the Mezo Portal so we may consider the address as trusted but I would consider adding reentrancy protection just to reduce the potential attack surface if things go wrong in the future (e.g. a malicious token is added by mistake).